### PR TITLE
support fair share

### PR DESCRIPTION
**Pull request by lminzhw**
Implementation of: https://github.com/volcano-sh/volcano/blob/master/docs/design/fairshare.md

support fair share based on weight among namespaces.

Part of: https://github.com/volcano-sh/volcano/issues/244

### CODE REVIEW

File: pkg/scheduler/api/namespace_info.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+const (
+	// QuotaKey is the key in ResourceQuota.spec.hard indicating the weight of thi namespace
+	QuotaKey = "volcano.sh/namespace.weight"
+	// DefaultWeight is the default weight of namespace
+	DefaultWeight = 1
+)
+
+// NamespaceInfo records information of namespace
+type NamespaceInfo struct {
+	Name string
```
Comment (hzxuzhonghu) [inline pkg/scheduler/api/namespace_info.go]: What's Name mean?

File: pkg/scheduler/api/namespace_info.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+// NewNamespaceCollection creates new NamespaceCollection object to record all information about a namespace
+func NewNamespaceCollection(name string) *NamespaceCollection {
+	n := &NamespaceCollection{
+		Name:        name,
+		quotaWeight: cache.NewHeap(quotaItemKeyFunc, quotaItemLessFunc),
+	}
+	// add at least one item into quotaWeight.
+	// Because cache.Heap.Pop would be blocked until queue is not empty
+	n.updateWeight(&quotaItem{
+		name:   QuotaKey,
```
Comment (hzxuzhonghu) [inline pkg/scheduler/api/namespace_info.go]: Is this necessary, since it is a fixed value.

File: pkg/scheduler/api/namespace_info.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	B := b.(*quotaItem)
+	return A.weight > B.weight
+}
+
+// NamespaceCollection will record all details about namespace
+type NamespaceCollection struct {
+	Name string
+
+	weightMu    sync.Mutex
+	quotaWeight *cache.Heap
```
Comment (hzxuzhonghu) [inline pkg/scheduler/api/namespace_info.go]: To my understanding, only need to keep the biggest weight.

Even if you want all the weights, there is a `PriorityQueue` in volcano, does it satisfy your requirement?

File: pkg/scheduler/api/namespace_info.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	A := a.(*quotaItem)
+	B := b.(*quotaItem)
+	return A.weight > B.weight
+}
+
+// NamespaceCollection will record all details about namespace
+type NamespaceCollection struct {
+	Name string
+
+	weightMu    sync.Mutex
```
Comment (hzxuzhonghu) [inline pkg/scheduler/api/namespace_info.go]: donnot need the lock

File: pkg/scheduler/actions/allocate/allocate.go [hunk 5]
```diff
@@ -43,7 +43,14 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
 	defer glog.V(3).Infof("Leaving Allocate ...")
 
 	queues := util.NewPriorityQueue(ssn.QueueOrderFn)
+	// map[queueId]PriorityQueue(*api.JobInfo)
 	jobsMap := map[api.QueueID]*util.PriorityQueue{}
+	// map[queueId]PriorityQueue(namespaceName)
+	namespaceMap := map[api.QueueID]*util.PriorityQueue{}
+	// map[queueId]map[namespaceName]PriorityQueue(*api.JobInfo)
+	jobInNamespaceMap := map[api.QueueID]map[string]*util.PriorityQueue{}
```
Comment (hzxuzhonghu) [inline pkg/scheduler/actions/allocate/allocate.go]: Add more comments

File: pkg/scheduler/api/cluster_info.go [hunk 6]
```diff
@@ -20,9 +20,10 @@ import "fmt"
 
 // ClusterInfo is a snapshot of cluster by cache.
 type ClusterInfo struct {
-	Jobs   map[JobID]*JobInfo
-	Nodes  map[string]*NodeInfo
-	Queues map[QueueID]*QueueInfo
+	Jobs          map[JobID]*JobInfo
+	Nodes         map[string]*NodeInfo
+	Queues        map[QueueID]*QueueInfo
+	NamespaceInfo map[string]*NamespaceInfo
```
Comment (hzxuzhonghu) [inline pkg/scheduler/api/cluster_info.go]: Add comments key-->value

File: pkg/scheduler/cache/event_handlers.go [hunk 7]
```diff
@@ -1129,3 +1129,87 @@ func (sc *SchedulerCache) addPriorityClass(pc *v1beta1.PriorityClass) {
...
+	sc.updateResourceQuota(newR)
+}
+
+// AddResourceQuota add ResourceQuota to scheduler cache
+func (sc *SchedulerCache) AddResourceQuota(obj interface{}) {
+	var r *v1.ResourceQuota
+	switch t := obj.(type) {
+	case *v1.ResourceQuota:
+		r = t
+	case cache.DeletedFinalStateUnknown:
```
Comment (hzxuzhonghu) [inline pkg/scheduler/cache/event_handlers.go]: donot need this in add

File: pkg/scheduler/cache/event_handlers.go [hunk 8]
```diff
@@ -1129,3 +1129,87 @@ func (sc *SchedulerCache) addPriorityClass(pc *v1beta1.PriorityClass) {
...
+
+	sc.Mutex.Lock()
+	defer sc.Mutex.Unlock()
+
+	glog.V(3).Infof("Update ResourceQuota <%s/%v> in cache, with spec: %v.", newR.Namespace, newR.Name, newR.Spec.Hard)
+	sc.updateResourceQuota(newR)
+}
+
+// AddResourceQuota add ResourceQuota to scheduler cache
+func (sc *SchedulerCache) AddResourceQuota(obj interface{}) {
```
Comment (hzxuzhonghu) [inline pkg/scheduler/cache/event_handlers.go]: I can not see where you filterout resource quota that does not set `vaolcano.sh/xx.weight`

File: pkg/scheduler/actions/allocate/allocate.go [hunk 9]
```diff
@@ -62,12 +69,37 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
...
+
+		jobs, found := namespaceJob[job.Namespace]
+		if !found {
+			jobs = util.NewPriorityQueue(ssn.JobOrderFn)
+			namespaceJob[job.Namespace] = jobs
+
+			namespaceMap[job.Queue].Push(job.Namespace)
+		}
+
+		glog.V(4).Infof("Added Job <%s/%s> into Queue <%s> Namespace <%s>", job.Namespace, job.Name, job.Queue, job.Namespace)
```
Comment (hzxuzhonghu) [inline pkg/scheduler/actions/allocate/allocate.go]: nit: duplicate namespace

File: pkg/scheduler/plugins/drf/drf.go [hunk 10]
```diff
@@ -83,18 +89,66 @@ func (drf *drfPlugin) OnSessionOpen(ssn *framework.Session) {
...
+			lNsAlloc := lNsAtt.allocated.Clone().Add(preemptor.Resreq)
+			_, lNsShare := drf.calculateShare(lNsAlloc, drf.totalResource)
+			lNsShareWeighted = lNsShare / float64(lWeight)
+		}
+
 		allocations := map[api.JobID]*api.Resource{}
+		namespaceAllocation := map[string]*api.Resource{}
 
 		for _, preemptee := range preemptees {
+			if namespaceOrderEnabled && preemptor.Namespace != preemptee.Namespace {
```
Comment (hzxuzhonghu) [inline pkg/scheduler/plugins/drf/drf.go]: more comments

File: pkg/scheduler/api/namespace_info.go (hunk 1 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: `Name` is the name of this namespace, I'll add comments above it.

File: pkg/scheduler/api/namespace_info.go (hunk 2 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: Now, `name` of `quotaItem` is used in `quotaItemKeyFunc` and it is for `cache.Heap` to distinguish different quota resources. 

`PriorityQueue` in volcano may have another behavior. I'll have a try.

File: pkg/scheduler/api/namespace_info.go (hunk 3 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: The only biggest weight will be kept in `NamespaceInfo` after `cache.Snapshot()`.

At here, `NamespaceCollection` store all the information about this namespace and will be able to handle the weight changes in `Quota` resource (such as try to decrease the biggest weight).

I'll have a look at `PriorityQueue` later~

Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: The `PriorityQueue` is a simple implementation of `container/heap`, but it could not support the `Update` action, and delete some item in heap directly is also not supported.

So, It it also the reason why I choose the Heap in k8s cache before.

File: pkg/scheduler/actions/allocate/allocate.go [hunk 11]
```diff
@@ -193,6 +253,10 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
 		} else {
 			stmt.Discard()
 		}
+
+		if namespaceOrderEnabled {
```
Comment (TommyLike) [inline pkg/scheduler/actions/allocate/allocate.go]: We would enter the place that we have a huge **Execute** method that is much too long 😄

File: pkg/scheduler/api/namespace_info_test.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
+package api
```
Comment (TommyLike) [inline pkg/scheduler/api/namespace_info_test.go]: Missing header here.

File: pkg/scheduler/api/namespace_info.go [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+// NamespaceInfo records information of namespace
+type NamespaceInfo struct {
+	Name string
+	// Weight is the highest weight among many ResourceQuota.
+	Weight int64
+}
+
+// GetWeight returns weight of a namespace, any invalid case would get default value
+func (n *NamespaceInfo) GetWeight() int64 {
+	if n == nil {
```
Comment (TommyLike) [inline pkg/scheduler/api/namespace_info.go]: too if statements can be combined.

File: pkg/scheduler/api/namespace_info.go (hunk 4 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: The original idea for me to add the lock is to guarantee the actions in `Snapshot` should be atomic.

Now, I also think this lock is unnecessary.

I'll remove this lock.

File: pkg/scheduler/api/namespace_info.go (hunk 2 above)
Comment (TommyLike) [reply pkg/scheduler/api/namespace_info.go]: Keep them all would be ok to me.

File: pkg/scheduler/cache/event_handlers.go [hunk 14]
```diff
@@ -1129,3 +1129,87 @@ func (sc *SchedulerCache) addPriorityClass(pc *v1beta1.PriorityClass) {
...
+func (sc *SchedulerCache) DeleteResourceQuota(obj interface{}) {
+	var r *v1.ResourceQuota
+	switch t := obj.(type) {
+	case *v1.ResourceQuota:
+		r = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		r, ok = t.Obj.(*v1.ResourceQuota)
+		if !ok {
+			glog.Errorf("Cannot convert to *v1beta1.PriorityClass: %v", t.Obj)
```
Comment (TommyLike) [inline pkg/scheduler/cache/event_handlers.go]: fix `v1beta1.PriorityClass`

File: pkg/scheduler/cache/event_handlers.go [hunk 15]
```diff
@@ -1129,3 +1129,87 @@ func (sc *SchedulerCache) addPriorityClass(pc *v1beta1.PriorityClass) {
...
+			return
+		}
+	default:
+		glog.Errorf("Cannot convert to *v1beta1.PriorityClass: %v", t)
```
Comment (TommyLike) [inline pkg/scheduler/cache/event_handlers.go]: ditto

File: pkg/scheduler/cache/event_handlers.go [hunk 16]
```diff
@@ -1129,3 +1129,87 @@ func (sc *SchedulerCache) addPriorityClass(pc *v1beta1.PriorityClass) {
...
+
+	glog.V(3).Infof("Delete ResourceQuota <%s/%v> in cache", r.Namespace, r.Name)
+	sc.deleteResourceQuota(r)
+}
+
+// UpdateResourceQuota update ResourceQuota to scheduler cache
+func (sc *SchedulerCache) UpdateResourceQuota(oldObj, newObj interface{}) {
+	newR, ok := newObj.(*v1.ResourceQuota)
+	if !ok {
+		glog.Errorf("Cannot convert newObj to *v1beta1.PriorityClass: %v", newObj)
```
Comment (TommyLike) [inline pkg/scheduler/cache/event_handlers.go]: ditto

File: pkg/scheduler/cache/event_handlers.go (hunk 8 above)
Comment (TommyLike) [reply pkg/scheduler/cache/event_handlers.go]: The namespace without this label would default to 1

File: pkg/scheduler/framework/session_plugins.go [hunk 17]
```diff
@@ -304,6 +309,50 @@ func (ssn *Session) JobOrderFn(l, r interface{}) bool {
...
+			if j := nof(l, r); j != 0 {
+				return j < 0
+			}
+		}
+	}
+
+	// TODO(lminzhw): if all NamespaceOrderFn treat these two namespace as the same,
+	// we should make the job order have its affect among namespaces.
+	// or just schedule namespace one by one
+	lv := l.(string)
```
Comment (TommyLike) [inline pkg/scheduler/framework/session_plugins.go]: Do we need to check if this converting failing?

File: pkg/scheduler/framework/session_plugins.go [hunk 18]
```diff
@@ -304,6 +309,50 @@ func (ssn *Session) JobOrderFn(l, r interface{}) bool {
...
+				continue
+			}
+			return isEnabled(plugin.EnabledNamespaceOrder)
+		}
+	}
+	return false
+}
+
+// NamespaceOrderFn invoke namespaceorder function of the plugins
+func (ssn *Session) NamespaceOrderFn(l, r interface{}) bool {
```
Comment (TommyLike) [inline pkg/scheduler/framework/session_plugins.go]: So the order of namespace is decided by the first plugin if it has enabled the namespace order?

File: pkg/scheduler/plugins/drf/drf.go (hunk 10 above)
Comment (TommyLike) [reply pkg/scheduler/plugins/drf/drf.go]: Better refactor this part. since the change is little too hardcoded.

File: pkg/scheduler/api/namespace_info.go [hunk 19]
```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (k82cn) [inline pkg/scheduler/api/namespace_info.go]: s/kubernetes/Volcano

Comment (k82cn) [reply pkg/scheduler/api/namespace_info.go]: new files belong to volcano community

File: pkg/scheduler/api/namespace_info.go [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	// QuotaKey is the key in ResourceQuota.spec.hard indicating the weight of thi namespace
+	QuotaKey = "volcano.sh/namespace.weight"
+	// DefaultWeight is the default weight of namespace
+	DefaultWeight = 1
```
Comment (k82cn) [inline pkg/scheduler/api/namespace_info.go]: default namespace weight

File: pkg/scheduler/api/namespace_info.go [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+
+	"github.com/golang/glog"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	// QuotaKey is the key in ResourceQuota.spec.hard indicating the weight of thi namespace
+	QuotaKey = "volcano.sh/namespace.weight"
```
Comment (k82cn) [inline pkg/scheduler/api/namespace_info.go]: s/QuotaKey/NamespaceWeightKey

File: pkg/scheduler/framework/session_plugins.go [hunk 22]
```diff
@@ -304,6 +309,50 @@ func (ssn *Session) JobOrderFn(l, r interface{}) bool {
...
 }
 
+// NamespaceOrderEnabled returns the NamespaceOrder is enabled in this session or not
+// judging by length of namespaceOrderFns
+func (ssn *Session) NamespaceOrderEnabled() bool {
+	return len(ssn.namespaceOrderFns) > 0
+}
+
+// NamespaceOrderEnabledInPlugin the NamespaceOrder for this plugin is enabled in this session or not
+func (ssn *Session) NamespaceOrderEnabledInPlugin(pluginName string) bool {
```
Comment (k82cn) [inline pkg/scheduler/framework/session_plugins.go]: move this to drf plugin; otherwise, there're lots of helper function in `Session`.

File: pkg/scheduler/plugins/drf/drf.go [hunk 23]
```diff
@@ -66,6 +70,8 @@ func (drf *drfPlugin) OnSessionOpen(ssn *framework.Session) {
 		drf.totalResource.Add(n.Allocatable)
 	}
 
+	namespaceOrderEnabled := ssn.NamespaceOrderEnabledInPlugin(drf.Name())
```
Comment (k82cn) [inline pkg/scheduler/plugins/drf/drf.go]: is there anyway to remove `namespaceOrderEnabled`? That's error prone :)

File: pkg/scheduler/actions/allocate/allocate.go (hunk 5 above)
Comment (lminzhw) [reply pkg/scheduler/actions/allocate/allocate.go]: add more comments to describe these two maps.

File: pkg/scheduler/api/cluster_info.go (hunk 6 above)
Comment (lminzhw) [reply pkg/scheduler/api/cluster_info.go]: add type `NamespaceName` to avoid too many comments

File: pkg/scheduler/cache/event_handlers.go (hunk 7 above)
Comment (lminzhw) [reply pkg/scheduler/cache/event_handlers.go]: delete this case block

File: pkg/scheduler/cache/event_handlers.go (hunk 8 above)
Comment (lminzhw) [reply pkg/scheduler/cache/event_handlers.go]: In `itemFromQuota` function, any `ResourceQuota` without `xx.weight` key would  have weight with default value 1

File: pkg/scheduler/actions/allocate/allocate.go (hunk 9 above)
Comment (lminzhw) [reply pkg/scheduler/actions/allocate/allocate.go]: change duplicate namespace log to prefix `NamespaceEnabled: ` to indicate this add action is different to adding job directly into map belonging to queue.

File: pkg/scheduler/plugins/drf/drf.go (hunk 10 above)
Comment (lminzhw) [reply pkg/scheduler/plugins/drf/drf.go]: refactor this part, namespace order policy is now a separate code block out of job order policy.

File: pkg/scheduler/api/namespace_info_test.go (hunk 12 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info_test.go]: add headers before package

File: pkg/scheduler/api/namespace_info.go (hunk 13 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: combine these two statements

File: pkg/scheduler/cache/event_handlers.go (hunk 14 above)
Comment (lminzhw) [reply pkg/scheduler/cache/event_handlers.go]: fix it using 'v1.ResourceQuota'

File: pkg/scheduler/cache/event_handlers.go (hunk 15 above)
Comment (lminzhw) [reply pkg/scheduler/cache/event_handlers.go]: fixed

File: pkg/scheduler/cache/event_handlers.go (hunk 16 above)
Comment (lminzhw) [reply pkg/scheduler/cache/event_handlers.go]: fixed

File: pkg/scheduler/framework/session_plugins.go (hunk 17 above)
Comment (lminzhw) [reply pkg/scheduler/framework/session_plugins.go]: this actions following the code pattern in other `CompareFn`, like `taskOrder` or `jobOrder` in `priority` plugin.

File: pkg/scheduler/framework/session_plugins.go (hunk 18 above)
Comment (lminzhw) [reply pkg/scheduler/framework/session_plugins.go]: Yes, if some plugin decide namespace A is prior to B, this decision is just the final result.

This logic follows the code pattern in other `CompareFn`, like `JobOrderFn`.

File: pkg/scheduler/api/namespace_info.go (hunk 19 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: fixed

File: pkg/scheduler/api/namespace_info.go (hunk 20 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: rename it to `DefaultNamespaceWeight`

File: pkg/scheduler/api/namespace_info.go (hunk 21 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: rename it

File: pkg/scheduler/framework/session_plugins.go (hunk 22 above)
Comment (lminzhw) [reply pkg/scheduler/framework/session_plugins.go]: move it to `drf.NamespaceOrderEnabled`.

File: pkg/scheduler/plugins/drf/drf.go (hunk 23 above)
Comment (lminzhw) [reply pkg/scheduler/plugins/drf/drf.go]: I refactor the code in allocate, there is no `namespaceOrderEnabled` in allocation func now.

And, `namespaceOrderEnabled` in `drf` plugin is also refactored to be more readable.

File: pkg/scheduler/api/namespace_info.go (hunk 1 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: move its type from string to `NamespaceName`

File: pkg/scheduler/actions/allocate/allocate.go (hunk 9 above)
Comment (lminzhw) [reply pkg/scheduler/actions/allocate/allocate.go]: after code refactoring, this log is deleted.

File: pkg/scheduler/actions/allocate/allocate.go (hunk 11 above)
Comment (lminzhw) [reply pkg/scheduler/actions/allocate/allocate.go]: refactor action Allocate, remove all `namespaceOrderEnabled` in it.

**Comment by lminzhw** [comment]
@hzxuzhonghu  @TommyLike  @k82cn 

TLDR: code and algorithm in `allocate.go` changes a lot.

---

In previous version, I use `PriorityQueue` to pick namespace from choosed queue.

But, when there is many queue, the changes of namespace priority by allocation in this queue would have no chance to affect  `PriorityQueue` in other queue.

So, I change the method to pick <namespace, queue> tuple.

---

And why I choose to pick namespace firstly but not queue?

we have two method to pick them:

- pick namespace firstly, with time complexity log(n) * q
- pick queue firstly, with time complexity log(q) * n

and the first picked one would have a more balanced resource usage.

In my opinion, the number of queue would less than namespace in most case, and the balanced resource usage of namespace should be considered first, so I choose to pick namespace firstly.

**Comment by lminzhw** [comment]
> I think we need more tests, especially for namespace job allocate part.
> 
> Also i noticed a problem: The namespce may not get requested resources even when there is enough quota. Say we have queue1 and queue2, ns1 and ns2, all with weight 1. Jobs in ns1 (requires 50% of all resources) are using queue1 and queue2. Jobs in ns2(requires 50% all resources) are using queue2 only.
> 
> If queue order is queue2 ->queue1, then ns2 can only get 25% of the resources. Is this correct?

After reading code carefully, that is not a problem now, I think.

In `propotion` plugin, the queue with less share would be allocation firstly. And in `allocate` action, after every job allocation, the queue with less share would have change to allocate pod.

So, if ns1 and ns2 have enough pod count (for having enough scheduling loop), queue2 be scheduled firstly or not is not that concerned, because the next scheduling loop would choose to schedule another for its lower share.

In newly pushed version, ns1 be scheduled firstly would lead to schedule ns2 next and the resource is also balanced.

File: pkg/scheduler/api/namespace_info.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/scheduler/api/namespace_info.go]: IC,  this is to support resourcequota update

**Comment by hzxuzhonghu** [comment]
LGTM, but i would expect e2e tests as well.

**Comment by lminzhw** [comment]
> LGTM, but i would expect e2e tests as well.

We have 2 more e2e test cases and a XXL PR now.  : )

**Comment by hzxuzhonghu** [comment]
@lminzhw Would suggest not squash all commits, it is hard to review.

File: test/e2e/job_scheduling.go [hunk 24]
```diff
@@ -387,4 +392,196 @@ var _ = Describe("Job E2E Test", func() {
 		err = waitJobPhaseReady(context, job)
 		Expect(err).NotTo(HaveOccurred())
 	})
+
```
Comment (hzxuzhonghu) [inline test/e2e/job_scheduling.go]: Better add a case with both queue and namespace fair share?

File: pkg/scheduler/api/namespace_info.go [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+	"fmt"
+
+	"github.com/golang/glog"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+// NamespaceName is name of namespace
+type NamespaceName string
```
Comment (hex108) [inline pkg/scheduler/api/namespace_info.go]: Would `Namespace` be other types other than `string`? If not, it might be better to use `string` directly. Then it avoids many type conversion

File: pkg/scheduler/actions/allocate/allocate.go [hunk 26]
```diff
@@ -91,21 +111,47 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
...
 		}
 
-		queue := queues.Pop().(*api.QueueInfo)
-		if ssn.Overused(queue) {
-			glog.V(3).Infof("Queue <%s> is overused, ignore it.", queue.Name)
-			continue
+		// pick namespace from namespaces PriorityQueue
+		namespace := namespaces.Pop().(api.NamespaceName)
+
+		queueInNamespace := jobsMap[namespace]
```
Comment (hex108) [inline pkg/scheduler/actions/allocate/allocate.go]: A little confused. What is the relationship between queue and namespace? According to the doc [motivation](https://github.com/volcano-sh/volcano/blob/master/docs/design/fairshare.md#motivation), a namespace works like a user, and there might be many users(namespaces) in a queue. Will be many queues in a namespace?

File: pkg/scheduler/actions/allocate/allocate.go [hunk 27]
```diff
@@ -91,21 +111,47 @@ func (alloc *allocateAction) Execute(ssn *framework.Session) {
...
+		// pick queue for given namespace
+		//
+		// This block use a algorithm with time complex O(n).
+		// But at least PriorityQueue could not be used here,
+		// because the allocation of job would change the priority of queue among all namespaces,
+		// and the PriorityQueue have no ability to update priority for a special queue.
+		var queue *api.QueueInfo
+		for queueId := range queueInNamespace {
+			currentQueue := ssn.Queues[queueId]
+			if ssn.Overused(currentQueue) {
```
Comment (hex108) [inline pkg/scheduler/actions/allocate/allocate.go]: If the queue is overused, could not the queue get resource even the cluster has idle resource?

File: pkg/scheduler/api/namespace_info.go [hunk 28]
```diff
@@ -0,0 +1,146 @@
...
+func quotaItemKeyFunc(obj interface{}) (string, error) {
+	item, ok := obj.(*quotaItem)
+	if !ok {
+		return "", fmt.Errorf("obj with type %T could not parse", obj)
+	}
+	return item.name, nil
+}
+
+// for big root heap
+func quotaItemLessFunc(a interface{}, b interface{}) bool {
```
Comment (hex108) [inline pkg/scheduler/api/namespace_info.go]: s/quotaItemLessFunc/quotaItemGreaterFunc/?

File: pkg/scheduler/plugins/drf/drf.go [hunk 29]
```diff
@@ -60,12 +64,27 @@ func (drf *drfPlugin) Name() string {
 	return PluginName
 }
 
+// NamespaceOrderEnabled returns the NamespaceOrder for this plugin is enabled in this session or not
+func (drf *drfPlugin) NamespaceOrderEnabled(ssn *framework.Session) bool {
```
Comment (hex108) [inline pkg/scheduler/plugins/drf/drf.go]: What is the difference between it and `func (ssn *Session) NamespaceOrderEnabled()`?

File: pkg/scheduler/actions/allocate/allocate.go (hunk 26 above)
Comment (Jeffwan) [reply pkg/scheduler/actions/allocate/allocate.go]: @hex108 As I understand,  generally speaking, different users (namespace) could schedule jobs in different Queues.  This doc I think concentrates more on the faire resource sharing in one queue between users. This solution is mainly trying to resolve starvation problem in resource sharing.

File: pkg/scheduler/api/namespace_info.go [hunk 30]
```diff
@@ -0,0 +1,146 @@
...
+	// DefaultNamespaceWeight is the default weight of namespace
+	DefaultNamespaceWeight = 1
+)
+
+// NamespaceInfo records information of namespace
+type NamespaceInfo struct {
+	// Name is the name of this namespace
+	Name NamespaceName
+	// Weight is the highest weight among many ResourceQuota.
+	Weight int64
```
Comment (Jeffwan) [inline pkg/scheduler/api/namespace_info.go]: if we use signed integer, what's the best practice here? Do we suggest users to always use weights > 0 or it could be negative numbers?

File: test/e2e/job_scheduling.go (hunk 24 above)
Comment (lminzhw) [reply test/e2e/job_scheduling.go]: This RP is too big now, I'll submit a new test next PR.

File: pkg/scheduler/api/namespace_info.go (hunk 25 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: I just want to make codes more readable, so choose to mark it as a new type but not mark it in comment.

File: pkg/scheduler/actions/allocate/allocate.go (hunk 27 above)
Comment (lminzhw) [reply pkg/scheduler/actions/allocate/allocate.go]: Yes, the overused queue will be skipped in master code too.

File: pkg/scheduler/api/namespace_info.go (hunk 28 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: done~

File: pkg/scheduler/plugins/drf/drf.go (hunk 29 above)
Comment (lminzhw) [reply pkg/scheduler/plugins/drf/drf.go]: The `NamespaceOrderEnabled` in plugin just return the value of that plugin. This func return the result of this session and used in `allocate` action.

And in the newest version code, I delete this func and will open a Issue to adjust the behavior when the NamespaceOrder is disabled or other cases.

File: pkg/scheduler/api/namespace_info.go (hunk 30 above)
Comment (lminzhw) [reply pkg/scheduler/api/namespace_info.go]: Yes, weight < 1 is meaningless in this case, because we have a default value which is 1. Any weight < 1 will be ignored.

**Comment by lminzhw** [comment]
@all, sorry for the delayed reply.

And this PR leaves two unsolved problem now:
1. add a case with both Queue and Namespace fair share
2. the behavior when disable NamespaceOrder is not so good, I will adjust it.

Waiting for your response~

**Comment by k82cn** [comment]
/lgtm

help to re-label lgtm, and it LGTM to :)
